server/src/main/java/org/opensearch/cluster/routing/allocation/allocator/BalancedShardsAllocator.java

```diff
@@ -745,11 +760,19 @@ public void moveShards() {
 
                 ShardRouting shardRouting = it.next();
 
+                // Ensure that replicas don't relocate if primaries are being throttled and primary first is enabled
```

Comment:
Isn't primary getting throttled specific to a shardRouting, couldn't catch how that applies to other shards

Reply:
If any primary shard is getting throttled for some reason, we want to prevent any replicas from moving away. Though, we can break the block instead of iterating over rest of the replica, since when movePrimaryFirst is true, all the primaries are returned first

Reply:
Maybe just add a debug log which would help identify why shards are NOT moving. Also wondering what would allocation explain return when primary is being throttled and replica is perceived as stuck.

Reply:
I have added log message to indicate why replicas are not moving.

1. Since decision is made based on primary shard throttling, allocation explain API that evaluates single shard cannot capture this. Explanation of movePrimaryFirst setting should document this behavior
2. I am not in favor of adding another setting and complicating things further. This setting is turned off by default and should be enabled only to enforce (not best effort) primary first movement.

Reply:
I guess we can put some logic in BSA#decideMove to return a THROTTLE decision for all eligible replicas which would otherwise return YES against canAllocate. Thoughts??

Reply:
As mentioned earlier that allocation explain API evaluates single shard and difficult for it to know whether primaries are getting throttled or not. Also replicas get throttled only when primaries do, hence the shard relocation should not look completely stuck when primaries are throttled and user will not probably use the allocation explain API. So, leaving it as is should be fine?